pypeit/core/datacube.py

```diff
@@ -0,0 +1,344 @@
+        relative wavelength dependent weights for each input pixel.
+    """
+    msgs.info("Calculating the optimal weights of each pixel")
+    # Determine number of files
+    if numfiles is None:
+        numfiles = np.unique(all_idx).size
+
+    # Generate a white light image of *all* data
+    msgs.info("Generating global white light image")
+    whitelight_Img, ivar = make_whitelight(all_ra, all_dec, all_wave, all_sci, all_wghts, np.zeros(all_ra.size),
+                                           dspat, numfiles=1)
+    whitelight_Img = whitelight_Img[:, :, 0]
+    # Find the location of the object with the highest S/N in the combined white light image
+    idx_max = np.unravel_index(np.argmax(whitelight_Img), whitelight_Img.shape)
+    msgs.info("Highest S/N object located at spaxel (x, y) = {0:d}, {1:d}".format(idx_max[0], idx_max[1]))
+
+    # Generate a master 2D WCS to register all frames
+    coord_min = [np.min(all_ra), np.min(all_dec), np.min(all_wave)]
+    coord_dlt = [dspat, dspat, dwv]
+    whitelightWCS = generate_masterWCS(coord_min, coord_dlt)
+    # Make the bin edges to be at +/- 1 pixels around the maximum (i.e. summing 9 pixels total)
+    numwav = int((np.max(all_wave) - np.min(all_wave)) / dwv)
+    xbins = np.array([idx_max[0]-1, idx_max[0]+2]) - 0.5
+    ybins = np.array([idx_max[1]-1, idx_max[1]+2]) - 0.5
+    spec_bins = np.arange(1 + numwav) - 0.5
+    bins = (xbins, ybins, spec_bins)
+
+    # Extract the spectrum of the highest S/N object
+    flux_stack = np.zeros((numwav, numfiles))
+    ivar_stack = np.zeros((numwav, numfiles))
+    for ff in range(numfiles):
+        msgs.info("Extracting spectrum of highest S/N detection from frame {0:d}/{1:d}".format(ff + 1, numfiles))
+        ww = (all_idx == ff)
+        # Extract the spectrum
+        pix_coord = whitelightWCS.wcs_world2pix(np.vstack((all_ra[ww], all_dec[ww], all_wave[ww] * 1.0E-10)).T, 0)
+        spec, edges = np.histogramdd(pix_coord, bins=bins, weights=all_sci[ww])
+        var, edges = np.histogramdd(pix_coord, bins=bins, weights=1/all_ivar[ww])
+        norm, edges = np.histogramdd(pix_coord, bins=bins)
+        normspec = (norm > 0) / (norm + (norm == 0))
+        var_spec = var[0, 0, :]
+        ivar_spec = (var_spec > 0) / (var_spec + (var_spec == 0))
+        # Calculate the S/N in a given spectral bin
+        flux_stack[:, ff] = spec[0, 0, :] * np.sqrt(normspec)  # Note: sqrt(nrmspec), is because we want the S/N in a _single_ pixel (i.e. not spectral bin)
+        ivar_stack[:, ff] = ivar_spec
+
+    mask_stack = (flux_stack != 0.0) & (ivar_stack != 0.0)
+    # Obtain a wavelength of each pixel
+    wcs_res = whitelightWCS.wcs_pix2world(np.vstack((np.zeros(numwav), np.zeros(numwav), np.arange(numwav))).T, 0)
+    wave_spec = wcs_res[:, 2] * 1.0E10
+    # Identify the reference spectrum (the one with the highest typical S/N ratio)
+    medsnr = np.median(flux_stack*np.sqrt(ivar_stack), axis=0)
```

Comment:
This seems redundant to me since the sn_weights code already does it if you don't set ref_spec.